Handle alerts in ocamldoc

Julow opened the pull request:
Follow-up to the discussion in https://github.com/ocaml/ocaml/pull/10983 about alerts and tags.

Tags and alerts are have intersecting jobs but only alerts are inspected by the compiler and eventually other tools. As @dbuenzli noted in https://github.com/ocaml/ocaml/pull/10983#issuecomment-1028170207, it is annoying to have to write the `deprecated` annotation for the documentation and for the compiler separately.
For example, the stdlib doesn't use the `@deprecated` tag.

This PR makes ocamldoc recognize alerts attached to values, which are rendered similarly to the deprecated tag:

**Alert foo.** Body of the alert.

The deprecated alert is lifted into the deprecated tag for better rendering and to avoid rendering it twice. `[@@deprecated]` and `[@@alert deprecated]` are treated as synonyms.

This is how the documentation of the stdlib changes:

<details>

```diff
diff --git a/apidoc_old/compilerlibref/Config.html b/apidoc_new/compilerlibref/Config.html
index 3664661bc..772f38788 100644
--- a/apidoc_old/compilerlibref/Config.html
+++ b/apidoc_new/compilerlibref/Config.html
@@ -136,13 +136,15 @@
 
 <pre><span id="VALocamlopt_cflags"><span class="keyword">val</span> ocamlopt_cflags</span> : <code class="type">string</code></pre><div class="info ">
 <div class="info-deprecated">
-<span class="warning">Deprecated. </span><a href="Config.html#VALocamlc_cflags"><code class="code"><span class="constructor">Config</span>.ocamlc_cflags</code></a> should be used instead.
+<span class="warning">Deprecated. </span>Use ocamlc_cflags instead.
+<a href="Config.html#VALocamlc_cflags"><code class="code"><span class="constructor">Config</span>.ocamlc_cflags</code></a> should be used instead.
     The flags ocamlopt should pass to the C compiler</div>
 </div>
 
 <pre><span id="VALocamlopt_cppflags"><span class="keyword">val</span> ocamlopt_cppflags</span> : <code class="type">string</code></pre><div class="info ">
 <div class="info-deprecated">
-<span class="warning">Deprecated. </span><a href="Config.html#VALocamlc_cppflags"><code class="code"><span class="constructor">Config</span>.ocamlc_cppflags</code></a> should be used instead.
+<span class="warning">Deprecated. </span>Use ocamlc_cppflags instead.
+<a href="Config.html#VALocamlc_cppflags"><code class="code"><span class="constructor">Config</span>.ocamlc_cppflags</code></a> should be used instead.
     The flags ocamlopt should pass to the C preprocessor</div>
 </div>
 
diff --git a/apidoc_old/compilerlibref/Load_path.html b/apidoc_new/compilerlibref/Load_path.html
index bf0f5293b..6919fba98 100644
--- a/apidoc_old/compilerlibref/Load_path.html
+++ b/apidoc_new/compilerlibref/Load_path.html
@@ -123,6 +123,8 @@
 
 <pre><span id="MODULEDir"><span class="keyword">module</span> <a href="Load_path.Dir.html">Dir</a></span>: <code class="code"><span class="keyword">sig</span></code> <a href="Load_path.Dir.html">..</a> <code class="code"><span class="keyword">end</span></code></pre>
 <pre><span id="VALadd"><span class="keyword">val</span> add</span> : <code class="type"><a href="Load_path.Dir.html#TYPEt">Dir.t</a> -&gt; unit</code></pre><div class="info ">
+<div class="info-deprecated">
+<span class="warning">Deprecated. </span></div>
 <div class="info-desc">
 <p>Old name for <a href="Load_path.html#VALappend_dir"><code class="code"><span class="constructor">Load_path</span>.append_dir</code></a></p>
 </div>
diff --git a/apidoc_old/compilerlibref/Longident.html b/apidoc_new/compilerlibref/Longident.html
index 1378a8ac8..a6439a1bb 100644
--- a/apidoc_old/compilerlibref/Longident.html
+++ b/apidoc_new/compilerlibref/Longident.html
@@ -111,6 +111,9 @@
 
 <pre><span id="VALlast"><span class="keyword">val</span> last</span> : <code class="type"><a href="Longident.html#TYPEt">t</a> -&gt; string</code></pre>
 <pre><span id="VALparse"><span class="keyword">val</span> parse</span> : <code class="type">string -&gt; <a href="Longident.html#TYPEt">t</a></code></pre><div class="info ">
+<div class="info-deprecated">
+<span class="warning">Deprecated. </span>this function may misparse its input,
+use "Parse.longident" or "Longident.unflatten"</div>
 <div class="info-desc">
 <p>This function is broken on identifiers that are not just "Word.Word.word";
    for example, it returns incorrect results on infix operators
diff --git a/apidoc_old/compilerlibref/index_values.html b/apidoc_new/compilerlibref/index_values.html
index a41464107..8d3d7cbf4 100644
--- a/apidoc_old/compilerlibref/index_values.html
+++ b/apidoc_new/compilerlibref/index_values.html
@@ -95,9 +95,9 @@
 </td></tr>
 <tr><td><a href="Load_path.html#VALadd">add</a> [<a href="Load_path.html">Load_path</a>]</td>
 <td><div class="info">
-<p>Old name for <a href="Load_path.html#VALappend_dir"><code class="code"><span class="constructor">Load_path</span>.append_dir</code></a></p>
+<span class="deprecated"><p>Old name for <a href="Load_path.html#VALappend_dir"><code class="code"><span class="constructor">Load_path</span>.append_dir</code></a></p>
 
-</div>
+</span></div>
 </td></tr>
 <tr><td><a href="Clflags.html#VALadd_arguments">add_arguments</a> [<a href="Clflags.html">Clflags</a>]</td>
 <td></td></tr>
@@ -2216,11 +2216,11 @@
 <td></td></tr>
 <tr><td><a href="Longident.html#VALparse">parse</a> [<a href="Longident.html">Longident</a>]</td>
 <td><div class="info">
-<p>This function is broken on identifiers that are not just "Word.Word.word";
+<span class="deprecated"><p>This function is broken on identifiers that are not just "Word.Word.word";
    for example, it returns incorrect results on infix operators
    and extended module paths.</p>
 
-</div>
+</span></div>
 </td></tr>
 <tr><td><a href="Warnings.html#VALparse_alert_option">parse_alert_option</a> [<a href="Warnings.html">Warnings</a>]</td>
 <td><div class="info">
diff --git a/apidoc_old/libref/Printexc.html b/apidoc_new/libref/Printexc.html
index 883e2a463..85601351e 100644
--- a/apidoc_old/libref/Printexc.html
+++ b/apidoc_new/libref/Printexc.html
@@ -142,6 +142,8 @@
 </div>
 
 <pre><span id="VALcatch"><span class="keyword">val</span> catch</span> : <code class="type">('a -&gt; 'b) -&gt; 'a -&gt; 'b</code></pre><div class="info ">
+<div class="info-deprecated">
+<span class="warning">Deprecated. </span>This function is no longer needed.</div>
 <div class="info-desc">
 <p><code class="code"><span class="constructor">Printexc</span>.catch&nbsp;fn&nbsp;x</code> is similar to <a href="Printexc.html#VALprint"><code class="code"><span class="constructor">Printexc</span>.print</code></a>, but
    aborts the program with exit code 2 after printing the
diff --git a/apidoc_old/libref/Printf.html b/apidoc_new/libref/Printf.html
index 204b24862..a923efb44 100644
--- a/apidoc_old/libref/Printf.html
+++ b/apidoc_new/libref/Printf.html
@@ -315,6 +315,8 @@
 <p>Deprecated</p>
 
 <pre><span id="VALkprintf"><span class="keyword">val</span> kprintf</span> : <code class="type">(string -&gt; 'b) -&gt; ('a, unit, string, 'b) <a href="Stdlib.html#TYPEformat4">format4</a> -&gt; 'a</code></pre><div class="info ">
+<div class="info-deprecated">
+<span class="warning">Deprecated. </span>Use Printf.ksprintf instead.</div>
 <div class="info-desc">
 <p>A deprecated synonym for <code class="code">ksprintf</code>.</p>
 </div>
diff --git a/apidoc_old/libref/Stdlib.Printexc.html b/apidoc_new/libref/Stdlib.Printexc.html
index e60e56b30..e6e78c7ff 100644
--- a/apidoc_old/libref/Stdlib.Printexc.html
+++ b/apidoc_new/libref/Stdlib.Printexc.html
@@ -137,6 +137,8 @@
 </div>
 
 <pre><span id="VALcatch"><span class="keyword">val</span> catch</span> : <code class="type">('a -&gt; 'b) -&gt; 'a -&gt; 'b</code></pre><div class="info ">
+<div class="info-deprecated">
+<span class="warning">Deprecated. </span>This function is no longer needed.</div>
 <div class="info-desc">
 <p><code class="code"><span class="constructor">Printexc</span>.catch&nbsp;fn&nbsp;x</code> is similar to <a href="Printexc.html#VALprint"><code class="code"><span class="constructor">Printexc</span>.print</code></a>, but
    aborts the program with exit code 2 after printing the
diff --git a/apidoc_old/libref/Stdlib.Printf.html b/apidoc_new/libref/Stdlib.Printf.html
index de5fb774d..9de67e75f 100644
--- a/apidoc_old/libref/Stdlib.Printf.html
+++ b/apidoc_new/libref/Stdlib.Printf.html
@@ -310,6 +310,8 @@
 <p>Deprecated</p>
 
 <pre><span id="VALkprintf"><span class="keyword">val</span> kprintf</span> : <code class="type">(string -&gt; 'b) -&gt; ('a, unit, string, 'b) <a href="Stdlib.html#TYPEformat4">format4</a> -&gt; 'a</code></pre><div class="info ">
+<div class="info-deprecated">
+<span class="warning">Deprecated. </span>Use Printf.ksprintf instead.</div>
 <div class="info-desc">
 <p>A deprecated synonym for <code class="code">ksprintf</code>.</p>
 </div>
diff --git a/apidoc_old/libref/Thread.html b/apidoc_new/libref/Thread.html
index ecbcaabe5..6a91b890c 100644
--- a/apidoc_old/libref/Thread.html
+++ b/apidoc_new/libref/Thread.html
@@ -157,6 +157,8 @@
 </div>
 
 <pre><span id="VALexit"><span class="keyword">val</span> exit</span> : <code class="type">unit -&gt; unit</code></pre><div class="info ">
+<div class="info-deprecated">
+<span class="warning">Deprecated. </span>Use 'raise Thread.Exit' instead.</div>
 <div class="info-desc">
 <p>Raise the <a href="Thread.html#EXCEPTIONExit"><code class="code"><span class="constructor">Thread</span>.<span class="constructor">Exit</span></code></a> exception.
     In a thread created by <a href="Thread.html#VALcreate"><code class="code"><span class="constructor">Thread</span>.create</code></a>, this will cause the thread
diff --git a/apidoc_old/libref/index_values.html b/apidoc_new/libref/index_values.html
index 7565b4da2..2dcc4cdc0 100644
--- a/apidoc_old/libref/index_values.html
+++ b/apidoc_new/libref/index_values.html
@@ -1624,11 +1624,11 @@
 </td></tr>
 <tr><td><a href="Printexc.html#VALcatch">catch</a> [<a href="Printexc.html">Printexc</a>]</td>
 <td><div class="info">
-<p><code class="code"><span class="constructor">Printexc</span>.catch&nbsp;fn&nbsp;x</code> is similar to <a href="Printexc.html#VALprint"><code class="code"><span class="constructor">Printexc</span>.print</code></a>, but
+<span class="deprecated"><p><code class="code"><span class="constructor">Printexc</span>.catch&nbsp;fn&nbsp;x</code> is similar to <a href="Printexc.html#VALprint"><code class="code"><span class="constructor">Printexc</span>.print</code></a>, but
    aborts the program with exit code 2 after printing the
    uncaught exception.</p>
 
-</div>
+</span></div>
 </td></tr>
 <tr><td><a href="Sys.html#VALcatch_break">catch_break</a> [<a href="Sys.html">Sys</a>]</td>
 <td><div class="info">
@@ -4046,9 +4046,9 @@
 </td></tr>
 <tr><td><a href="Thread.html#VALexit">exit</a> [<a href="Thread.html">Thread</a>]</td>
 <td><div class="info">
-<p>Raise the <a href="Thread.html#EXCEPTIONExit"><code class="code"><span class="constructor">Thread</span>.<span class="constructor">Exit</span></code></a> exception.</p>
+<span class="deprecated"><p>Raise the <a href="Thread.html#EXCEPTIONExit"><code class="code"><span class="constructor">Thread</span>.<span class="constructor">Exit</span></code></a> exception.</p>
 
-</div>
+</span></div>
 </td></tr>
 <tr><td><a href="Stdlib.html#VALexit">exit</a> [<a href="Stdlib.html">Stdlib</a>]</td>
 <td><div class="info">
@@ -8732,9 +8732,9 @@
 </td></tr>
 <tr><td><a href="Printf.html#VALkprintf">kprintf</a> [<a href="Printf.html">Printf</a>]</td>
 <td><div class="info">
-<p>A deprecated synonym for <code class="code">ksprintf</code>.</p>
+<span class="deprecated"><p>A deprecated synonym for <code class="code">ksprintf</code>.</p>
 
-</div>
+</span></div>
 </td></tr>
 <tr><td><a href="Scanf.html#VALkscanf">kscanf</a> [<a href="Scanf.html">Scanf</a>]</td>
 <td><div class="info">
```

</details>

Octachron committed:
ocamldoc: cleanup info merging

Julow committed:
Test case for alerts

Julow committed:
ocamldoc: Handle @@alert and @@deprecated attributes

[@@deprecated] is treated the same as [@@alert deprecated]. Alerts are
rendered one per line, at the end of the comment.

Julow committed:
ocamldoc: Pass alerts along with 'info'

Fits better with the way the 'info' record is passed around and
rendered. Fix alerts not being rendered when there's not also a comment,
with the previous approach.

Julow committed:
ocamldoc: Unify the deprecated tag and alert

Remove the deprecated alert if the tag is present. Lift the alert into
the tag otherwise.

Julow committed:
ocamldoc: Support alerts in other backends

And render a '.' after the alert name. This is how the deprecated tag is
rendered. We expect a sentence in the payload.

Octachron commented on ocamldoc/odoc_messages.ml:
```diff
@@ -200,6 +200,7 @@ let merge_param = ('p', "merge @param")
 let merge_raised_exception = ('e', "merge @raise")
 let merge_return_value = ('r', "merge @return")
 let merge_custom = ('c', "merge custom @-tags")
+let merge_alert = ('x', "merge alerts")
```
'x' for `alerts`?

Julow replied:
Every letters of `alerts` are taken except `t` and I thought alerts are some kind of next-gen tags and needed a fancy letter.

Octachron replied:
Fine with me.

Octachron commented on ocamldoc/odoc_ast.ml:
```diff
@@ -1137,6 +1139,7 @@ module Analyser =
               else
                 None
             in
+            let comment_opt = Odoc_sig.analyze_alerts comment_opt val_desc.Parsetree.pval_attributes in
```
That's clever and probably the best way to source attributes information from the AST while keeping the separated extraction of documentation comments.

Octachron commented:
In term of features,  that's look like a good idea (even more so if that avoid delay a little the divergence of tags/attributes support between ocamldoc and odoc). 
I will try to fully review the code soon.

Octachron commented on ocamldoc/odoc_html.ml:
```diff
@@ -2063,10 +2071,10 @@ class html =
-           self#html_of_info_first_sentence
-        ) b m.m_info
+           self#html_of_info_first_sentence b m.m_info
+        )
```
Would you mind moving those style changes into their own commit?

Julow replied:
This was only needed for an intermediate patch that is reverted. So I reverted this change.

Octachron commented on ocamldoc/odoc_man.ml:
```diff
@@ -227,8 +244,9 @@ class virtual info =
-                (self#str_man_of_custom info.M.i_custom)
+              ]
+              @ self#str_man_of_custom info.M.i_custom
+              @ self#str_man_of_alerts info.M.i_alerts
```
It is probably better to have `custom` at the end of the info section in all backends.

Octachron commented on ocamldoc/odoc_misc.ml:
```diff
@@ -197,6 +197,15 @@ let string_of_return_opt return_opt =
+    (fun al ->
+      let payload = Option.to_list al.Odoc_types.alert_payload in
+      Odoc_messages.alert :: " " :: al.Odoc_types.alert_name :: " "
+      :: (payload @ [ "\n" ]))
```
There is a empty space before `\n` when the payload is empty.

Octachron commented on ocamldoc/odoc_sig.ml:
```diff
@@ -774,6 +830,7 @@ module Analyser =
             in
             let name = Name.parens_if_infix name_pre.txt in
             let subst_typ = Odoc_env.subst_type env type_expr in
+            let comment_opt = analyze_alerts comment_opt value_desc.Parsetree.pval_attributes in
```
This is too early to merge alert and deprecation information: at this point, we have not yet collected the post-element documentation comments. Thus if we transfer the deprecation alerts to the deprecation info, when calling
```ocaml
 merge_infos v.val_info info_after_opt
```
later we end up merging the alert and the deprecation comment.
This is what is happening in the html test.

Julow replied:
Thanks for the debugging ! I think I've found a fix.

Octachron commented on testsuite/tests/tool-ocamldoc/Alerts.html.reference:
```diff
@@ -0,0 +1,78 @@
+<pre><span id="VALg"><span class="keyword">val</span> g</span> : <code class="type">int</code></pre><div class="info ">
+<div class="info-deprecated">
+<span class="warning">Deprecated. </span>foo
+bar</div>
```
We have
```
<span class="warning">Deprecated. </span>foo
bar</div>
```
where I expected

```
<span class="warning">Deprecated. </span>foo</div>
```
See the comment above.

Julow replied:
I don't know how I could miss this! Perhaps some premature refactoring.

Octachron commented on ocamldoc/odoc_to_text.ml:
```diff
@@ -144,6 +144,18 @@ class virtual info =
+            | Some p -> [ Raw ". "; Raw p ]
+            | None -> []
+          in
+          [ Bold [ Raw Odoc_messages.alert; Raw al.alert_name ] ]
```
You need a white space between `Alert` and the alert name here.

Julow replied:
Fixed.

Julow committed:
ocamldoc: Render alerts before custom tags

Julow committed:
ocamldoc: Fix merging of deprecated tag and alert messages

Julow committed:
Update Changes

Julow committed:
ocamldoc: Remove unused info_string_of_info

This function is no longer uptodate but can be removed instead of fixed.

Julow commented:
Thanks for the review. Everything is fixed.

Octachron commented:
Sorry, there is a small feature gap that I had missed earlier: we want to handle compilation unit-wide deprecation and alerts:
```ocaml
(* a.mli *)
  [@@@deprecated "Dont' use" ]
```
Not handling deprecation on constructors or fields in perfectly fine for ocamldoc, but we probably want to handle that compilation unit case.

Julow committed:
ocamldoc: Handle unit-wide alerts

Look for signature-item attributes ([@@@...]) at the beginning of a
signature. These alerts are attached to the whole unit, similarly to the
preamble.

Julow committed:
ocamldoc: Handle alerts attached to modules and module types

Alerts attached to the module declaration and alerts at the beginning of
the signature are handled (even though a comment at this position
wouldn't be picked as the preamble):

    module M : sig
      [@@@deprecated "foo"]
    end

    module N : sig
    end
    [@@deprecated "foo"]

s module types

Julow committed:
ocamldoc: Remove some code duplication

Julow commented:
I just added support for unit-wide alerts and also for alerts attached to modules and module types (also "signature wide" alerts).
I didn't realize alerts could be attached to fields and constructors. I'll handle that now (same for types, exceptions, etc..)

Octachron commented:
Like I said, those cases are probably less important. But I will gladly review the code if you want to support them.

Julow committed:
ocamldoc: Handle attributes in other items

Constructors, fields and all signature items.

Julow commented:
Added a commit for this.

Octachron commented on ocamldoc/odoc_sig.ml:
```diff
@@ -1211,6 +1212,18 @@ module Analyser =
+            let comment_opt = analyze_alerts comment_opt pmd_attributes in
+            let comment_opt =
+              match module_type.Parsetree.pmty_desc with
+              | Parsetree.Pmty_signature sg -> analyze_toplevel_alerts comment_opt sg
```
Toplevel alerts are only recognized by the compiler at the compilation-unit level:
```ocaml
module M :sig [@@@deprecated] end
```
does not attach the alert to `M` currently.
This is already slightly confusing, I would rather not have `ocamldoc` add more confusion here.

Julow replied:
I've pushed a fix.

Julow committed:
ocamldoc: Don't extract alerts from signatures

The compiler doesn't do that, so this is inconsistent.

Octachron commented on testsuite/tests/tool-ocamldoc/Alert_toplevel2.mli:
```diff
@@ -0,0 +1,5 @@
+   * ocamldoc with html
+*)
+
+[@@@alert deprecated "foo"]
```
I would propose to keep the same docstring comment that the `Alert_toplevel.mli` test in order to have exactly the same reference file.

Julow replied:
That's what I wanted to test actually, that the docstring doesn't interfere with the alert. That it's using a different kind of alert is a bonus. (it should be covered already in the other tests because it's using the same code)

Octachron commented on ocamldoc/odoc_args.ml:
```diff
@@ -137,6 +137,7 @@ let analyse_merge_options s =
     (M.merge_raised_exception, [Odoc_types.Merge_raised_exception]) ;
     (M.merge_return_value, [Odoc_types.Merge_return_value]) ;
     (M.merge_custom, [Odoc_types.Merge_custom]) ;
+    (M.merge_alert, [Odoc_types.Merge_alert]) ;
```
Thinking a bit about this point, this is not the right semantic for alerts: alerts from the implementation are never merged nor lifted into the interface. This is not an user controlled behavior on the compiler side and thus it should not be an option on the ocamldoc side.

Julow replied:
Unrelated to this PR but isn't it weird that documentation comments aren't hidden by an interface file ? The generated documentation has more content than what's in the `mli` file and some documentation might be incomplete when switching from ocamldoc to odoc, which doesn't look at the implementation if it founds an interface.

Octachron replied:
This is part of the behavior of ocamldoc that was intentionally not reproduced in odoc, because merging information from two slightly incompatible different sources can be really confusing (as illustrated by the `@param` tag).

Julow replied:
I tweaked the merge but a bug remain: The deprecated alert is converted to a deprecated tag too early and will be merged anyway.
This is was on purpose initially but I'm happy to change that if you think we should. It would be a big change though.

Octachron replied:
No, that's enough time spent on ocamldoc. At this point, it is more a bug on the ocamldoc implementation of deprecation tags.

Octachron commented on ocamldoc/odoc_ast.ml:
```diff
@@ -1078,7 +1079,7 @@ module Analyser =
               [] ->
                 (acc_env, acc)
-            | {Parsetree.pvb_pat=pat; pvb_expr=exp} :: q ->
+            | {Parsetree.pvb_pat=pat; pvb_expr=exp;pvb_attributes=attrs} :: q ->
```
The behavior for single ml is now desynchronized from the behavior for interface file since we are only lifting alerts on value to the documentation side.
It is fine to not handle implementation only alerts, but it seems better to not partially handle them.

Julow replied:
Added the feature to the implem frontend too.

Octachron replied:
Great! Thanks a lot for all the time spent on those details.

Julow committed:
ocamldoc: Handle alerts in implementation

Julow committed:
ocamldoc: Don't merge alerts

And be sure that alerts from an implementation are not rendered if an
interface is present.

Octachron reviewed (approved):
Independently of #10983, surfacing all alert attributes in the generated documentation is a good change. In particular, this PR makes makes it much easier to keep the deprecation message synchronized between the alert and the documentation.
